strawberry/schema/schema_converter.py

```diff
@@ -348,37 +368,38 @@ def _strawberry_info_from_graphql(info: GraphQLResolveInfo) -> Info:
                 path=info.path,
             )
 
-        def _resolver(_source: Any, info: GraphQLResolveInfo, **kwargs):
-            strawberry_info = _strawberry_info_from_graphql(info)
-            _check_permissions(_source, strawberry_info, kwargs)
-
-            args, kwargs = _get_arguments(
-                source=_source, info=strawberry_info, kwargs=kwargs
+        def _get_result(_source: Any, info: Info, **kwargs):
+            field_args, field_kwargs = _get_arguments(
+                source=_source, info=info, kwargs=kwargs
             )
 
-            result = field.get_result(
-                _source, info=strawberry_info, args=args, kwargs=kwargs
+            return field.get_result(
+                _source, info=info, args=field_args, kwargs=field_kwargs
             )
 
-            if isasyncgen(result):
-
-                async def yield_results(results):
-                    async for value in results:
-                        yield value
+        def _resolver(_source: Any, info: GraphQLResolveInfo, **kwargs):
+            strawberry_info = _strawberry_info_from_graphql(info)
+            _check_permissions(_source, strawberry_info, kwargs)
 
-                return yield_results(result)
+            return _get_result(_source, strawberry_info, **kwargs)
 
-            if iscoroutine(result):  # pragma: no cover
+        async def _async_resolver(_source: Any, info: GraphQLResolveInfo, **kwargs):
+            strawberry_info = _strawberry_info_from_graphql(info)
+            await _check_permissions_async(_source, strawberry_info, kwargs)
 
-                async def await_result(result):
-                    return await result
+            result = _get_result(_source, strawberry_info, **kwargs)
 
-                return await_result(result)
+            if inspect.isawaitable(result):
+                return await result
 
             return result
 
-        _resolver._is_default = not field.base_resolver  # type: ignore
-        return _resolver
+        if field.has_async_permission_classes:
```

Comment:
```suggestion
        if field.has_async_permission_classes or (field.base_resolver is not None and field.base_resolver.is_async):
```

Reply:
Might be nice to have an `is_async` property on the field that encompasses all this logic.

Reply:
Excellent idea, didn't saw base_resolver had a similar property already.